File: find-eventual-safe-states/find-eventual-safe-states.java
class Solution {
    public List<Integer> eventualSafeNodes(int[][] graph) {
        boolean[] rec = new boolean[graph.length];
        boolean[] vis = new boolean[graph.length];
        List<Integer> res = new ArrayList<>();
        for (int i = 0; i < graph.length; i++) 
        {
            if(!hasCycle(graph, i, vis, rec))
                res.add(i);
        }
        return res;
    }
    private boolean hasCycle(int[][] graph, int i, boolean[] vis, boolean[] rec) {
        
        if (rec[i]) 
            return true;
        if (vis[i]) 
            return false;
        vis[i]=true;
        rec[i] = true;
        for (int next : graph[i]) 
        {
            if(hasCycle(graph, next, vis, rec)) 
                return true;
        }
        rec[i] = false;
        return false;
    }
}

File: 820-find-eventual-safe-states/find-eventual-safe-states.py
class Solution:
    def eventualSafeNodes(self, graph: List[List[int]]) -> List[int]:
        reverse_graph = defaultdict(list)
        for i in range(len(graph)):
            for edge in graph[i]:
                reverse_graph[edge].append(i)

        indegree = [0] * len(graph)
        for i in range(len(graph)):
            for node in reverse_graph[i]:
                indegree[node] += 1
        queue = deque([n for n in range(len(indegree)) if indegree[n] == 0])
        res = []
        while queue:
            node = queue.popleft()
            res.append(node)
            for edge in reverse_graph[node]:
                indegree[edge] -= 1
                if indegree[edge] == 0:
                    queue.append(edge)
        return sorted(res)

